stan/math/gpu/kernels/matrix_multiply.hpp

```diff
@@ -0,0 +1,89 @@
+#ifndef STAN_MATH_GPU_KERNELS_MATRIX_MULTIPLY_HPP
+#define STAN_MATH_GPU_KERNELS_MATRIX_MULTIPLY_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/gpu/kernel_cl.hpp>
+
+namespace stan {
+namespace math {
+namespace opencl_kernels {
+// \cond
+const char* matrix_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Matrix multiplication on the GPU
+     *
+     * @param[in] A the left matrix in matrix multiplication
+     * @param[in] B the right matrix in matrix multiplication
+     * @param[out] C the output matrix
+     * @param[in] M Number of rows for matrix A
+     * @param[in] N Number of rows for matrix B
+     * @param[in] K Number of cols for matrix A and number of rows for matrix B
+     */
+    __kernel void matrix_multiply(
+        const __global read_only double* A, const __global read_only double* B,
+        __global write_only double* C, const read_only int M,
+        const read_only int N, const read_only int K) {
+      // workitem index inside the workgroup
+      const int workgroup_row = get_local_id(0);
+      const int workgroup_col = get_local_id(1);
```

Comment:
Should these the the workitem rows/cols?

Also should we change all the ints, unsigned ints, etc. that we use in the kernels for getting indices into `size_t`?

Reply:
See previous comment.